modules/dnstap/dnstap.c

```diff
@@ -0,0 +1,410 @@
+	*has_port = true;
+
+	if (sockaddr->sa_family == AF_INET) {
+		const struct sockaddr_in *sai;
+		sai = (const struct sockaddr_in *)sockaddr;
+		addr->len = sizeof(sai->sin_addr);
+		addr->data = (uint8_t *)&sai->sin_addr.s_addr;
+		*port = ntohs(sai->sin_port);
+	} else if (sockaddr->sa_family == AF_INET6) {
+		const struct sockaddr_in6 *sai6;
+		sai6 = (const struct sockaddr_in6 *)sockaddr;
+		addr->len = sizeof(sai6->sin6_addr);
+		addr->data = (uint8_t *)&sai6->sin6_addr.s6_addr;
+		*port = ntohs(sai6->sin6_port);
+	}
+}
+
+/* dnstap_log prepares dnstap message and sent it to fstrm */
+static int dnstap_log(kr_layer_t *ctx) {
+	struct kr_request *req = ctx->req;
+	struct kr_module *module = ctx->api->data;
+	struct kr_rplan *rplan = &req->rplan;
+	struct dnstap_data *dnstap_dt = module->data;
+
+	/* check if we have a valid iothread */
+	if (!dnstap_dt->iothread || !dnstap_dt->ioq ) {
+		DEBUG_MSG("dnstap_dt->iothread or dnstap_dt->ioq is NULL\n");
+		return kr_error(EFAULT);
+	}
+
+	/* current time */
+	struct timeval now;
+	gettimeofday(&now, NULL);
+
+	/* Create dnstap message */
+	Dnstap__Message m;
+
+	memset(&m, 0, sizeof(m));
+
+	m.base.descriptor = &dnstap__message__descriptor;
+	/* Only handling response */
+	m.type = DNSTAP__MESSAGE__TYPE__RESOLVER_RESPONSE;
+
+	if (req->qsource.tcp) {
+		m.socket_protocol = DNSTAP__SOCKET_PROTOCOL__TCP;
+	} else {
+		m.socket_protocol = DNSTAP__SOCKET_PROTOCOL__UDP;
+	}
+	m.has_socket_protocol = true;
+
+	if (req->qsource.addr) {
+		set_address(req->qsource.addr,
+				&m.query_address,
+				&m.has_query_address,
+				&m.query_port,
+				&m.has_query_port);
+	}
+
+	if (req->qsource.dst_addr) {
+		set_address(req->qsource.dst_addr,
+				&m.response_address,
+				&m.has_response_address,
+				&m.response_port,
+				&m.has_response_port);
+		switch (req->qsource.dst_addr->sa_family) {
+			case AF_INET:
+				m.socket_family = DNSTAP__SOCKET_FAMILY__INET;
+				m.has_socket_family = true;
+				break;
+			case AF_INET6:
+				m.socket_family = DNSTAP__SOCKET_FAMILY__INET6;
+				m.has_socket_family = true;
+				break;
+		}
+	}
+
+	if (dnstap_dt->log_resp_pkt) {
+		const knot_pkt_t *rpkt = req->answer;
+		m.response_message.len = rpkt->size;
+		m.response_message.data = (uint8_t *)rpkt->wire;
+		m.has_response_message = true;
+	}
+
+	/* set query time to the timestamp of the first kr_query
+	 * set response time to now
+	 */
+	if (rplan->resolved.len > 0 ) {
+		struct kr_query *first = rplan->resolved.at[0];
+
+		m.query_time_sec = first->timestamp.tv_sec;
+		m.has_query_time_sec = true;
+		m.query_time_nsec = first->timestamp.tv_sec * 1000;
+		m.has_query_time_nsec = true;
+	}
+
+	/* Response time */
+	m.response_time_sec = now.tv_sec;
+	m.has_response_time_sec = true;
+	m.response_time_nsec = now.tv_sec * 1000;
+	m.has_response_time_nsec = true;
+
+	/* Qname */
+	m.query_zone.data = (uint8_t *)knot_pkt_qname(req->answer);
+	m.query_zone.len = req->answer->qname_size;
+	m.has_query_zone = true;
```

Comment:
`query_zone` should be recovered from current zone cut, like here: https://github.com/CZ-NIC/knot-resolver/blob/master/lib/resolve.c#L1021